cylc/flow/scripts/completion_server.py

```diff
@@ -0,0 +1,686 @@
+            Max time to wait for new requests.
+
+    Note:
+        The server will not play nicely with completions which contain spaces
+        (they will show as separate completions).
+
+        Completions should not contain spaces, workflows/cycles/tasks/jobs
+        cannot contain spaces (due to validation rules) so this is ok.
+
+    """
+    for line in listener(timeout=timeout):
+        try:
+            write(' '.join(
+                item
+                for item in sorted(await responder(*line[:-1].split('|')))
+            ))
+        except Exception:
+            # catch everything! If an error occurs swallow it, we don't want
+            # it to reach the CLI or to cause the server to get caught in
+            # a crash/restart loop
+            write('')  # send an empty reply
+        if once:
+            return
+        sys.stdout.flush()
+
+
+async def complete_cylc(_root: str, *items: str) -> t.List[str]:
+    """Return Cylc CLI completions for the provided args.
+
+    Args:
+        _root:
+            The root command i.e. `cylc`.
+        items:
+            The command items e.g. ['trigger', 'workfl']
+
+    Returns:
+        List of possible completions e.g. ['workflow_a//', 'workflow_b//']
+
+    """
+    length: int = len(items)
+
+    # the word to complete
+    partial: t.Optional[str] = None
+    if length > 0:
+        partial = items[-1]
+
+    # the previous word
+    previous: t.Optional[str] = None
+    if length > 1:
+        previous = items[-2]
+
+    # complete the cylc sub-command
+    if length == 0:
+        return await complete_command()
+    if length == 1:
+        return await complete_command(partial)
+    command: str = items[0]
+
+    # special logic for the pseudo "help" and "version" commands
+    if command == 'help':
+        return complete(partial, ['id', 'all'])
+    if command == 'version':
+        return complete(partial, ['--long'])
+
+    # complete --options
+    if partial and partial.startswith('-'):
+        ret = await complete_option(command, partial)
+        if ret is not None:
+            return ret
+    if previous and previous.startswith('-'):
+        ret = await complete_option_value(command, previous, partial)
+        if ret is not None:
+            return ret
+
+    # complete positional arguments
+    return await complete_argument(command, partial)
+
+
+def complete(
+    partial: t.Optional[str],
+    words: t.Iterable[str]
+) -> t.List[str]:
+    """Return only matching words.
+
+    Args:
+        partial:
+            The string to match.
+        words:
+            The list of words to match against.
+
+    Returns:
+        The filtered list of matching words.
+
+    Examples:
+        >>> complete('w', ['beans', 'walrus', 'wellington'])
+        ['walrus', 'wellington']
+
+    """
+    if partial is None:
+        return list(words)
+    return [
+        word
+        for word in words
+        if word.startswith(partial)
+    ]
```

Comment:
This is preventing the inferred-run completions getting through. 

For example, `partial foo//`  now correctly infers `words foo/run4//2, ...`, but they get dropped here because they don't start with `foo//` (two `/`s) 

Interestingly if change line 231 to ` if word.startswith(partial[:-1])` the first TAB results in `foo// ---> foo/run4//` which implies the partial arg is actually replaced rather than appended to by the completion. If so, I suppose we could just make the run number explicit here (if the user types `foo//<TAB>`) and take the explicit name for subsequent completion steps (in which case your new infer_run code is currently doing some unnecessary work by finding cycle points?). Hopefully that makes sense :face_with_spiral_eyes:  

:arrow_up: makes it work, and makes the run number explicit - which I think is a good thing.  i.e. `cylc blah foo//<TAB> ---> cylc blah foo/run4//`

Reply:
When we infer the run number we must remove it from the results returned.